### PR TITLE
Use generics for sharedcomponents, removes casting increases type safetiness

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

File: receiver/otlpreceiver/factory.go [hunk 1]
```diff
@@ -119,4 +122,4 @@ func createLog(
 // create separate objects, they must use one otlpReceiver object per configuration.
 // When the receiver is shutdown it should be removed from this map so the same configuration
 // can be recreated successfully.
-var receivers = sharedcomponent.NewSharedComponents()
+var receivers = sharedcomponent.NewSharedComponents[*Config, *otlpReceiver]()
```
Comment (djaglowski) [inline receiver/otlpreceiver/factory.go:125]: 👍

**Review by djaglowski — APPROVED** [review]
LGTM

**Review by codeboten — APPROVED** [review]
Please rebase

**Pull request merged**